Machine_Learning/notebooks/czmodel/Regresssion_PyTorch_5_0_0.ipynb

```diff
@@ -0,0 +1,676 @@
+    "This part fits the model to the loaded data. In this test example we do not care about an actual evaluation of the model using validation and test datasets."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define the number of training epochs\n",
+    "NUM_EPOCHS = 30\n",
+    "\n",
+    "# Define loss function and optimizer\n",
+    "loss_fn = torch.nn.MSELoss()\n",
+    "optimizer = torch.optim.Adam(model.parameters(), lr=1e-3, eps=1e-08)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_one_epoch(epoch_index):\n",
+    "    running_loss = 0.\n",
+    "\n",
+    "    for i, data in enumerate(sample_dataloader):\n",
+    "        inputs, labels = data\n",
+    "        inputs = inputs.to(device)\n",
+    "        labels = labels.to(device)\n",
+    "        optimizer.zero_grad()\n",
+    "        outputs = model(inputs)\n",
+    "\n",
+    "        loss = loss_fn(outputs, labels)\n",
+    "        loss.backward()\n",
+    "\n",
+    "        optimizer.step()\n",
+    "        running_loss += loss.item()\n",
+    "\n",
+    "    return running_loss"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "99kOKKv4tJOK"
+   },
+   "outputs": [],
+   "source": [
+    "def plot_history(loss):\n",
+    "    plt.plot(loss)\n",
+    "    plt.title('training loss')\n",
+    "    plt.ylabel('loss')\n",
+    "    plt.xlabel('epoch')\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 1000
+    },
+    "id": "hY5nMzGVrEEG",
+    "outputId": "81245272-1cca-4812-ce96-e966f40d909d",
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "training_loss = []\n",
+    "for epoch in range( NUM_EPOCHS):\n",
+    "    loss = train_one_epoch(epoch)\n",
+    "    training_loss.append(loss)\n",
+    "    print(f'Epoch {epoch+1}, loss %2e'%(loss))\n",
+    "\n",
+    "# Plot the training history\n",
+    "plot_history(training_loss)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 309
+    },
+    "id": "cgU2oygWyBVC",
+    "outputId": "080cdba2-66c0-4263-cd83-af1a1cd74c12"
+   },
+   "outputs": [],
+   "source": [
+    "# Plot the training results\n",
+    "images_labels  = next(iter(sample_dataloader))\n",
+    "image, label = images_labels\n",
+    "\n",
+    "plt.figure(figsize=(16,16))\n",
+    "# Plot the inputs\n",
+    "plt.subplot(1,3,1)\n",
+    "plt.axis('off')\n",
+    "plt.title('Inputs')\n",
+    "plt.imshow(image[0, 0, ...], cmap='gray')\n",
+    "\n",
+    "# Plot the predictions\n",
+    "prediction = model(image.to(device))\n",
```

Comment:
Where is _device_ defined?

![image](https://user-images.githubusercontent.com/9250590/190163423-286dca95-e757-4f65-896d-08c9a6d19e0e.png)